SQLite DB with SQLAlchemy II

yedpodtrzitko opened the pull request:
This takes some parts which has been done in #190 and keep working on it further

The transfer from JSON backend to SQL backend isnt as simple as replacing `json.load()` with `db.select()` so there is a lot of more refactoring to do before this will be production ready.  The current JSON implementation suffers with memleaks (see #208) and is slightly over-engineered at some places, so that's yet another reason for a refactoring.

The reason why even open the PR at this time is to keep track of what's done, and to avoid potential duplication of work in case someone else would feel like working on the DB implementation. In such case, feel free to open PR against this branch.

Unfortunately I dont keep updating this branch with anything what's landing in the `main` branch, so by that time this will be any close to being done, the `main` branch might divert very far.


## Migration process
- [x] opening library
- [x] displaying items
- [x] pagination
- [x] removing entry fields
- [x] adding entry fields
- [x] creating tags 
- [x] items selection 
- [x] searching via tag name
- [x] searching with include/exclude list 
- [x] reserve tag IDs under 1000 for internal use
- [x] bug: searching is now case-sensitive
- [x] save paths in POSIX format
- [x] bug: Archived/Favorite tags are misbehaving with multiselect
- [x] bug: title field in multiselect shows as "Mixed Data"
- [x] feat: Macros - Folders to Tags
- [x] feat: Macros - Sort Fields
- [x] feat: Macros - Autofill
- [x] feat: Unlinked Entries
- [x] feat: move fields into DB
- [x] Closing Library errors
- [x] bug: Closing library keeps pagination shown
- [x] feat: Save Library Backup
- [x] misc: Save Library (probably just remove, all operations go directly to DB)
- [ ] feat: subtags/parent tags
- [ ] feat: tag alias
- [ ] feat: tag shorthand

CyanVoxel commented:
This looks like a great barebones(ish) replacement to the JSON system so far, I think this is the best bet to get a SQL in the door without worrying about new schemas.

Taking a look at your comments, I'm keeping track of the features that still need to be implemented. So far I can tell there's at least:
- Select All functionality
- Various Macro functionality (was unofficial in the first place so not a priority)
- Ability to purge items from the `NavigationState`/`FilterState`
- Probably everything else that hasn't been touched?

From my own precursory testing, when opening an existing library with 200 items (191 visible with extension list filter), it comes back with supposedly 20 invisible results and tells me that most if not all the items are "not in library yet", such as with the following log snippet:
```
2024-07-24 23:49:35 [info     ] item not in library yet        path=WindowsPath('3D/d20 scene.mtl')
2024-07-24 23:49:35 [info     ] item not in library yet        path=WindowsPath('3D/d20 scene.obj')
```
Refreshing the library for new files also does not change anything here. Could be a case of the Windows paths acting funky somewhere, although that's just a surface-level guess.

I would also recommend basing this against the [Alpha-v9.4](https://github.com/TagStudioDev/TagStudio/tree/Alpha-v9.4) branch whenever it feels right, that way the changes there don't come as a sudden surprise. I don't *think* there's much going on there that will get in the way of this, but new changes are new changes nonetheless. I don't expect this to be ready by 9.4, but if it were then I would give this precedent over any new feature PRs. I feel that this would likely be included in 9.5 along with the new search syntax additions (now *that* will need some work resolving). 

Please let me know what I can do to help speed this along or help smooth the transition in any way as well. Thank you so much for working on this!

yedpodtrzitko commented:
@CyanVoxel I originally planned to polish this a bit more before getting it merged, but looking how the project got a second breath now, and a lot of new PRs are getting created, it  would be better to get merged as soon as possible, so any new PRs would be based against the changes here. I'm not THAT much invested in this to be backporting every single PR into this branch. Working on this functionality makes me busy enough already, and not every single PR can be easily merged, if at all.

For example this PR #329 from the Alpha-9.4 branch for sorting tags - the sorting happens manually, whereas with DB you can just offload the sorting to the SQL query. Merging this PR would mean... I cant really merge it, because the underlying parts are now totally different.

Same goes with other library related PRs like #284 or #322  - the proposed changes are not possible to get merged, as they basically operate on a dictionary structure (ie. the loaded json) instead of the database.

Same applies with non-library PRs, for example #235 - it says "logging refactoring", but it's actually needlessly touching way more things, which will create a lot of git conflicts when merging that. And the list keeps going...

So even though not everything here is 100% finished, having it merged could speed up getting things done, as it wouldnt rely just on me, but other people could fix some of the issues and/or add required functionality to the database backend.

Regarding your previous comment - most of the problems you mentioned are already fixed, as a lot of changes landed here since then. I have also added some actually meaningful tests, and the tests are now testing the QT application as well, so that could help quite a lot to not introduce any regressions, and have well defined behaviour.

CyanVoxel commented:
> @CyanVoxel I originally planned to polish this a bit more before getting it merged, but looking how the project got a second breath now, and a lot of new PRs are getting created, it would be better to get merged as soon as possible, so any new PRs would be based against the changes here. I'm not THAT much invested in this to be backporting every single PR into this branch. Working on this functionality makes me busy enough already, and not every single PR can be easily merged, if at all.
> 
> For example this PR #329 from the Alpha-9.4 branch for sorting tags - the sorting happens manually, whereas with DB you can just offload the sorting to the SQL query. Merging this PR would mean... I cant really merge it, because the underlying parts are now totally different.
> 
> Same goes with other library related PRs like #284 or #322 - the proposed changes are not possible to get merged, as they basically operate on a dictionary structure (ie. the loaded json) instead of the database.
> 
> Same applies with non-library PRs, for example #235 - it says "logging refactoring", but it's actually needlessly touching way more things, which will create a lot of git conflicts when merging that. And the list keeps going...
> 
> So even though not everything here is 100% finished, having it merged could speed up getting things done, as it wouldnt rely just on me, but other people could fix some of the issues and/or add required functionality to the database backend.
> 
> Regarding your previous comment - most of the problems you mentioned are already fixed, as a lot of changes landed here since then. I have also added some actually meaningful tests, and the tests are now testing the QT application as well, so that could help quite a lot to not introduce any regressions, and have well defined behaviour.

First off, thank you *so much* for working so hard on this. This is insanely valuable to the project and it'll be a game changer to finally be working off of SQL!

This PR is my `#1` priority for this project as far as I'm concerned. I'll take a look and see how things are operating now and get this merged into `main` as long as nothing critical in your migration list is broken. In the meantime, I'll probably do some sort of preview release for 9.4 based off of the current state of the `thumbnails` branch to get those features out in some form, and then divert my resources to helping get this polished off and have the 9.4 features backported to work with this over time.

seakrueger commented on tagstudio/src/core/constants.py:
```diff
@@ -124,45 +127,10 @@
+
+
+class LibraryPrefs(Enum):
+    IS_EXCLUDE_LIST = False
+    EXTENSION_LIST: list[str] = []
```
This is inverted from the current release. Current release is exclude by default with ".json", ".xmp", ".aae"

seakrueger commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,769 @@
+            relative_path = path.relative_to(self.library_dir)
+            # TODO - load these in batch somehow
+            if not self.has_item(path=relative_path):
+                logger.info("item not in library yet", path=relative_path)
```
This generates a lot of log spam when creating a library. Could this (and other messages) be moved to debug and just a summary logged to info?

seakrueger commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,769 @@
+                    session.add(Preferences(key=pref.name, value=pref.value))
+                    session.commit()
+                except IntegrityError:
+                    logger.error("pref already exists", pref=pref)
```
This errors out twice everytime a library is opened. Would is make more sense to not have an error message like adding the default tags above? or adding some sort of check to see if they default stuff is missing?

seakrueger commented on tagstudio/src/core/library/alchemy/fields.py:
```diff
@@ -0,0 +1,311 @@
+from sqlalchemy import ForeignKey, UniqueConstraint
+from sqlalchemy.orm import Mapped, mapped_column, relationship
+
+from .db import Base
```
Sorry a very minor nit. Is there a reason implicit imports are used for alchemy directory?

yedpodtrzitko replied:
what implicit import? Do you mean relative import? Probably autocompleted by PyCharm like this.

seakrueger replied:
yeah sorry, relative import

yedpodtrzitko replied:
afaik there isnt any coding standard rule about imports. When there will be, it would be better to enforce it via ruff or whatever too will be used. So I'll leave it as is for now.

seakrueger reviewed (commented):
A dependency on libX11 is also added, I'm not entirely sure where, but probably from one of the newly packages? I don't think its a huge deal though...

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/enums.py:
```diff
@@ -0,0 +1,88 @@
+    """Represent a state of the Library grid view."""
+
+    page_index: int = 0
+    page_size: int = 100
```
I dont have a library with 500+ items, so in order to test pagination I needed a lower number here. This should be ideally configurable through the `Preferences` table.

yedpodtrzitko commented:
@seakrueger thanks for the review, appreciate it.

CyanVoxel commented:
I've been giving this a spin and taking notes, thank you again so much for working on this!
I won't pester with non-critical stuff for this PR. Just a few questions and observations of relevant things to address before deciding to pull:

1. I noticed that new Tag IDs are starting after the internal ones instead of at 1000. Those IDs were previously reserved for required built-in tags, with "Favorite" and "Archived" being the only current ones. That way the program would have room to keep track of tags that need special treatment in some way via a consistently known ID. I wanted to get your take on that, whether we should continue with ID reservation or devise a more apt solution for the future, such as a dedicated table for required built-in tags. This is also likely going to be an issue with exportable/importable tags in the future, however fine-tuning these solutions is *far* out of the scope of what we're doing here. But there will need to be a decision on the fate of those reserved IDs before users are given the opportunity to populate them.
**TL;DR:** I don't think it would hurt to reserve those IDs again now and contemplate alternative solutions later.
2. I see that every entry now has a "Meta Tags" field by default. Was this due to a requirement by the database (see 5.), or was this done in an effort to quicken the tagging process by having a premade field? If it's the later, I say let's double down and offer "Title", "Meta Tags", and "Content Tag" fields at least. A future PR can make this user-configurable. 
3. Paths are currently being stored in the DB as OS-specific, however due to some nonsense this can lead to the issue described in #298. Part of this fix was going to be always storing the paths as POSIX inside the DB going forward. The rest of that fix involved converting the paths in existing libraries, but that can be saved as part of the JSON to SQL conversion process.
4. It doesn't seem like subtags are implemented yet, correct?
5. Lastly I wanted to note that the Favorite + Archived badges act very out of sync with the entries when in multiselection and when manually adding/removing the Favorite and Archived tags or fields containing them from entries, even after refreshing the pages. The strangest behavior I've been able to replicate is deleting the premade Meta Tags field, clicking the favorite button, and not getting the Favorite tag to show until adding the Meta Tags field back in. Anyway, I don't mind if badges aren't working for this PR, I just wanted to make sure that there weren't any deeper DB issues related to it.

Thank you so much again for all this!

yedpodtrzitko commented:
>     1. I noticed that new Tag IDs are starting after the internal ones instead of at 1000. 

will add an item into the checklist, thanks for bringing this to my attention

> 
>     2. I see that every entry now has a "Meta Tags" field by default. Was this due to a requirement by the database (see 5.), or was this done in an effort to quicken the tagging process by having a premade field? If it's the later, I say let's double down and offer "Title", "Meta Tags", and "Content Tag" fields at least. A future PR can make this user-configurable.

I think it was just because when I wanted to test anything related to tags, I had to add a new field each time, which gets old pretty fast. I can either remove that, or add all the other fields you mentioned.


>     3. Paths are currently being stored in the DB as OS-specific

added into the checklist


>     4. It doesn't seem like subtags are implemented yet, correct?

db structure is prepared for both parent tags and subtags, but I havent tried to actually create any, so I cant confirm nor deny this.

>     5. Lastly I wanted to note that the Favorite + Archived badges act very out of sync with the entries when in multiselection and when manually adding/removing the Favorite and Archived tags or fields containing them from entries, even after refreshing the pages. The strangest behavior I've been able to replicate is deleting the premade Meta Tags field, clicking the favorite button, and not getting the Favorite tag to show until adding the Meta Tags field back in. Anyway, I don't mind if badges aren't working for this PR, I just wanted to make sure that there weren't any deeper DB issues related to it.

added into checklist

CyanVoxel commented:
> > [...]
> 
> will add an item into the checklist, thanks for bringing this to my attention
> 
> > [...]
>
> I think it was just because when I wanted to test anything related to tags, I had to add a new field each time, which gets old pretty fast. I can either remove that, or add all the other fields you mentioned.
> 
> > [...]
>
> added into the checklist
> 
> > [...]
>
> db structure is prepared for both parent tags and subtags, but I havent tried to actually create any, so I cant confirm nor deny this.
> 
> > [...]
> 
> added into checklist

Thank you for the response! For the addition of default fields, I say let's go ahead with that. Probably more convenient than having none, and again this can be made user-configurable in a future PR. Everything else sounds good! 👍

eivl commented on tagstudio/src/core/library/alchemy/fields.py:
```diff
@@ -0,0 +1,173 @@
+    DATE_PUBLISHED = DefaultField(
+        id=14, name="Date Published", type=FieldTypeEnum.DATETIME
+    )
+    # ARCHIVED = DefaultField(id=15, name="Archived",  type=CheckboxField.checkbox)
+    # FAVORITE = DefaultField(id=16, name="Favorite", type=CheckboxField.checkbox)
```
Are this not needed?

yedpodtrzitko replied:
I didnt find it used anywhere; it was probably used in past before `Favorite` and `Archived` was changed to a tag, but I want to keep it around just in case

eivl replied:
fair!

eivl reviewed (commented):
Reviewd everything up to this point in time and made comments along the way.

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/db.py:
```diff
@@ -0,0 +1,48 @@
+    Base.metadata.create_all(engine)
+
+    # tag IDs < 1000 are reserved
+    # create tag and delete it to bump the autoincrement sequence
```
the table `sqlite_sequence` doesnt exist at this moment for some reason. I didnt want to get stuck on that, so for now I do this. If you can figure it out, I'd be happy to have a cleaner solution here.

eivl replied:
Im happy to look into it as its own issue once this massive work you have started is done!

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/joins.py:
I dont see which line you're commenting on

eivl replied:
Me neither.... guess i was tired. :D ignore this.

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,771 @@
+        color=TagColor.YELLOW,
+    )
+
+    return [archive_tag, favorite_tag]
```
I probably needed it mutable at some point in the past and then left it like this even after the reason for that vanished.

yedpodtrzitko commented on tagstudio/tests/qt/test_driver.py:
```diff
@@ -0,0 +1,62 @@
+
+
+def test_update_thumbs(qt_driver):
+    qt_driver.frame_content = [Entry(path=Path("/tmp/foo"))]
```
I dont think the test ever touches filesystem, so there shouldnt be anything breaking here. 

What other platforms than Windows might misbehave?

Also see #298 - the proposed solution is to always save paths in posix format.

eivl replied:
some unix like system where there are no `/tmp` folder. even with posix format, the folder `c:/tmp/foo/` does not exist. 

Windows however handles both `\` `\\` and `/` gracefully.

yedpodtrzitko replied:
The test doesnt touch filesystem - it just creates db records which needs to have some value in the property `.path`, but no actual files are created.

yedpodtrzitko commented on tagstudio/src/qt/widgets/tag_box.py:
```diff
@@ -127,62 +122,60 @@ def set_tags(self, tags: list[int]):
+            done_callback=self.driver.preview_panel.update_widgets,
             has_save=True,
         )
         # self.edit_modal.widget.update_display_name.connect(lambda t: self.edit_modal.title_widget.setText(t))
```
it sure can, as thousands of other lines... but I dont want to make the diff bigger than it already is.

eivl replied:
Thats fair, we can also automate the process of removing this in its own PR.

yedpodtrzitko commented on tagstudio/src/qt/widgets/tag.py:
```diff
@@ -13,15 +12,10 @@
-
 class TagWidget(QWidget):
     edit_icon_128: Image.Image = Image.open(
         str(Path(__file__).parents[3] / "resources/qt/images/edit_icon_128.png")
```
I didnt touch this line within this PR, so I consider that out of scope. But there's new `ResourceManager` which should eventually handle all resources like this one:

https://github.com/TagStudioDev/TagStudio/blob/bc38e568fa3da18e7a05c2931472a98a0e9ac2bc/tagstudio/src/qt/resource_manager.py#L14

yedpodtrzitko commented on tagstudio/src/qt/widgets/preview_panel.py:
```diff
@@ -644,63 +660,38 @@ def update_widgets(self):
+            for i, grid_idx in enumerate(self.driver.selected[1:]):
+                item = self.driver.frame_content[grid_idx]
+                item_field_types = {f.type for f in item.fields}
+                for f in common_fields:
+                    if f.type not in item_field_types:
+                        common_fields.remove(f)
+                        mixed_fields.append(f)
+
```
good catch

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,771 @@
+        Ensure that the default builtin tags  are present in the Library's
+        save file. Takes in and returns the tag dictionary from the JSON file.
+        """
+        missing: list = []
+
+        for m in missing:
+            tag_list.append(m)
+
+        return tag_list
```
I dont think this is even used, just copy-paste from the json library class. Gonna remove it.

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,771 @@
+        self.ignored_extensions = []
+        self.missing_matches = {}
+
+    def refresh_dir(self) -> Iterator[int]:
```
I havent touched this method yet, it's copy-paste from json library.

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,771 @@
+        with Session(self.engine) as session:
+            # add all items
+            session.add_all(items)
+            session.flush()
```
`autoflush` parameter is `True` by default, but the behaviour is different than when I flush explicitly

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,771 @@
+            full_path = self.library_dir / entry.path
+            if not full_path.exists() or not full_path.is_file():
+                self.missing_files.append(str(full_path))
+            yield i
```
copy-paste from json library; havent touched this yet

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,850 @@
+            elif search.id:
+                statement = statement.where(Entry.id == search.id)
+            elif search.tag_id:
+                # TODO
```
right now probably just to remove the forgotten `TODO` comment :)

eivl replied:
fantatsic! :D

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,850 @@
+        field: LibraryField,
+        entry_ids: list[int],
+    ) -> None:
+        # TODO - optimize
```
I'm not done with it yet, I just needed to make it work while working on other things, but I want to make it better without repeating myself that much. So the TODO here is appropriate I'd say.

eivl replied:
I agree, i just imagined i saw a checkbox beside your long list if things to do in the main post. very much appropriate!

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/library.py:
```diff
@@ -0,0 +1,850 @@
+        field: LibraryField,
+        content: str | datetime.datetime | set[Tag],
+        entry_ids: list[int],
+        # mode: Literal["replace", "append", "remove"],
```
it was in the original code, but afaik it has been never implemented so for now it can be probably removed.

yedpodtrzitko commented on tagstudio/src/core/palette.py:
```diff
@@ -1,274 +1,281 @@
+from src.core.library.alchemy.enums import TagColor
 
-class ColorType(int, Enum):
+logger = structlog.get_logger(__name__)
```
for now I just wanted to focus on the actual database than niceties around. Using `logging` is cumbersome so I replaced it with `structlog` but I dont want to think where and how to have logging placed and configured, so I use this approach for now. This can sure be refactored out later.

yedpodtrzitko commented on tagstudio/src/core/library/alchemy/db.py:
```diff
@@ -0,0 +1,48 @@
+from sqlalchemy import Dialect, Engine, String, TypeDecorator, create_engine, text
+from sqlalchemy.orm import DeclarativeBase
+
+logger = structlog.getLogger(__name__)
```
for now I just want to focus on the actual database rather than niceties around. Using default `logging` is cumbersome so I replaced it with `structlog`, but I dont want to think where and how to have the logging placed and configured, so I use this approach for now. This can sure be refactored out later.


> how would structlog handle threading

Since it shows the logs from other threads by default, I dont really dig deep into it


> And for windows you also need colorama for colors; and Rich or better-exceptions if you want more helpful output.

I'm happy with the default exceptions, as that will be better when eventually someone will be copy-pasting the traceback when reporting a bug.

eivl commented on tagstudio/src/qt/main_window.py:
```diff
@@ -36,7 +36,7 @@ class Ui_MainWindow(QMainWindow):
   
     def __init__(self, driver: "QtDriver", parent=None) -> None:
```
```suggestion
    def __init__(self, driver: QtDriver, parent=None) -> None:
```

yedpodtrzitko replied:
this is not a line I have touched, so I consider that out of scope.

eivl commented on tagstudio/src/qt/modals/delete_unlinked.py:
```diff
@@ -28,10 +28,10 @@
     done = Signal()
 
-    def __init__(self, library: "Library", driver: "QtDriver"):
+    def __init__(self, driver: "QtDriver", tracker: MissingRegistry):
```
though, this and possible the rest of the places, you would need to import the class for this type. 

```suggestion
    def __init__(self, driver: QtDriver, tracker: MissingRegistry):
```

eivl reviewed (approved):
ignore my previous comments, im just so used to doing 
```python
from __future__ import annotations
```
so I did not notice it was just a typing thing.

yedpodtrzitko commented:
@CyanVoxel apart from the last three items in the checklist, everything else should be done. I presume the tag-related items wont need any big changes, so it can be eventually done in `main` branch as well, so I'd say it's ready to get it merged to unblock merging other things, and more people can get their hands on it to test everything.

CyanVoxel commented:
> @CyanVoxel apart from the last three items in the checklist, everything else should be done. I presume the tag-related items wont need any big changes, so it can be eventually done in `main` branch as well, so I'd say it's ready to get it merged to unblock merging other things, and more people can get their hands on it to test everything.

Awesome!! Thank you so much for this again!

CyanVoxel commented:
I'm having some new difficulties with the imports since the last time I pulled this. I've made sure to reinstall the dependencies via `requirements.txt` and `requirements-dev.txt`. Have any of the dependencies or versions changed in the past couple weeks? Or may I be missing something obvious?

Windows 10, VSCode:
![image](https://github.com/user-attachments/assets/be3ead15-0128-4244-8903-85e58daf2f08)

Running `pip install -r requirements.txt`:
`Requirement already satisfied: greenlet!=0.4.17 in f:\github\tagstudio\.venv\lib\site-packages (from SQLAlchemy==2.0.34->-r requirements.txt (line 14)) (3.0.3)`

Running `pip install greenlet`:
`Requirement already satisfied: greenlet in f:\github\tagstudio\.venv\lib\site-packages (3.0.3)`

Running `pip list`:
```
Package                   Version
------------------------- --------
greenlet                  None
```
Along with the rest of the modules, with a few others having version "None" as well.

Computerdores commented:
the dependencies are not an issue for me (I ran my pip install commands after switching branch though), however Closing the library is broken for me it says "Closing & Saving Library", but doesn't actually do that

```
Traceback (most recent call last):
  File "C:\Users\Jann\Downloads\TagStudio\TagStudio\src\qt\ts_qt.py", line 326, in <lambda>
    close_library_action.triggered.connect(lambda: self.close_library())
                                                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jann\Downloads\TagStudio\TagStudio\src\qt\ts_qt.py", line 579, in close_library
    self.lib.clear_internal_vars()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Library' object has no attribute 'clear_internal_vars'
```

yedpodtrzitko commented:
> I'm having some new difficulties with the imports since the last time I pulled this. I've made sure to reinstall the dependencies via `requirements.txt` and `requirements-dev.txt`. Have any of the dependencies or versions changed in the past couple weeks? Or may I be missing something obvious?

`sqlalchemy` version was increased from `2.0.32` to `2.0.34`, which shouldnt make difference, but try installing that older version to see if that would help? `pip install sqlalchemy==2.0.32`

You can also try uninstall and reinstall `greenlet`:

```
pip uninstall greenlet
pip install -U greenlet
```


I dont have `greenlet` installed in my virtualenv at all, so this is quite surprising error to see.

yedpodtrzitko commented:
> the dependencies are not an issue for me (I ran my pip install commands after switching branch though), however closing the library is broken for me:
> 
> ```
> Traceback (most recent call last):
>   File "C:\Users\Jann\Downloads\TagStudio\TagStudio\src\qt\ts_qt.py", line 326, in <lambda>
>     close_library_action.triggered.connect(lambda: self.close_library())
>                                                    ^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\Jann\Downloads\TagStudio\TagStudio\src\qt\ts_qt.py", line 579, in close_library
>     self.lib.clear_internal_vars()
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'Library' object has no attribute 'clear_internal_vars'
> ```

fixed, thanks.

CyanVoxel commented:
> `sqlalchemy` version was increased from `2.0.32` to `2.0.34`, which shouldnt make difference, but try installing that older version to see if that would help? `pip install sqlalchemy==2.0.32`
> 
> You can also try uninstall and reinstall `greenlet`:
> 
> ```
> pip uninstall greenlet
> pip install -U greenlet
> ```
> 
> I dont have `greenlet` installed in my virtualenv at all, so this is quite surprising error to see.

Recreating the venv seems to have fixed it, and using either sqlalchemy 2.0.32 and 2.0.34 seems to work fine. Greenlet now lists the version as 3.0.3 in `pip list` 👍

CyanVoxel commented:
So far I've come across a few errors or situations where the data acts strangely. I'm not sure how many of these you're already aware of, but I'll share them here regardless:

Error upon loading a SQLite library populated with tags applied to entries:
<details><summary markdown="span">Show Logs</summary>
<pre>
2024-09-06 20:05:37 [info     ] update_widgets                 selected=[]
2024-09-06 20:05:37 [info     ] opening library                connection_string=sqlite:///F:\Example SQL\.TagStudio\ts_library.sqlite
2024-09-06 20:05:37 [info     ] creating db tables
2024-09-06 20:05:38 [info     ] update_widgets                 selected=[]
2024-09-06 20:05:38 [info     ] searching library              filter=FilterState(page_index=0, page_size=500, search_mode=<SearchMode.AND: 0>, tag=None, tag_id=None, id=None, path=None, name=None, query=None) query_full=SELECT entries.id, entries.path 
FROM entries
WHERE lower(entries.path) NOT LIKE lower('%..json,.xmp,.aae')
LIMIT 500 OFFSET 0
2024-09-06 20:05:38 [info     ] items to render                count=232
2024-09-06 20:05:38 [info     ] toggle_item_tag                entry_id=4 field_key=TAGS_META tag_id=1 toggle_value=True
2024-09-06 20:05:38 [error    ] Can't attach instance <Tag at 0x23f62791370>; another instance with key (<class 'src.core.library.alchemy.models.Tag'>, (1,), None) is already present in this session.
Traceback (most recent call last):
  File "F:\GitHub\TagStudio/tagstudio\src\core\library\alchemy\library.py", line 641, in add_field_tag
    field.tags = field.tags | {tag}
    ^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 537, in __set__
    self.impl.set(
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 1985, in set
    collections.bulk_replace(
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 804, in bulk_replace
    appender(member, _sa_initiator=initiator)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 1394, in add
    value = __set(self, value, _sa_initiator, NO_KEY)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 1096, in __set
    item = executor.fire_append_event(item, _sa_initiator, key=key)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 688, in fire_append_event
    return self.attr.fire_append_event(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 1752, in fire_append_event
    value = fn(state, value, initiator or self._append_token, key=key)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\unitofwork.py", line 70, in append
    sess._save_or_update_state(item_state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 3501, in _save_or_update_state
    self._save_or_update_impl(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 4202, in _save_or_update_impl
    self._update_impl(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 4191, in _update_impl
    self.identity_map.add(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\identity.py", line 195, in add
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Can't attach instance <Tag at 0x23f62791370>; another instance with key (<class 'src.core.library.alchemy.models.Tag'>, (1,), None) is already present in this session.
2024-09-06 20:05:38 [info     ] update_widgets                 selected=[]
2024-09-06 20:05:38 [info     ] searching library              filter=FilterState(page_index=0, page_size=500, search_mode=<SearchMode.AND: 0>, tag=None, tag_id=4, id=4, path=None, name=None, query=None) query_full=SELECT entries.id, entries.path 
</pre>
</details>


Error upon adding a tag to an entry, in certain circumstances? I'm aware that subtags are not yet implemented, but I'm not sure if that's the direct cause of this error. I created a tag "Cat", then  a tag "Kitten" with "Cat" as a subtag (didn't expect it to have any lasting effect), then applied "Cat" to an entry under content tags, then attempted to add "Kitten" as well, but failed. Creating a new tag and applying it to the same field worked as intended. Applying tag "Kitten" to "Meta Tags" also worked as intended. 
<details><summary markdown="span">Show Logs</summary>
<pre>
2024-09-06 20:18:10 [info     ] add_tag_callback               selected=[4] tag_id=1001
2024-09-06 20:18:10 [error    ] Can't attach instance <Tag at 0x21280572300>; another instance with key (<class 'src.core.library.alchemy.models.Tag'>, (1000,), None) is already present in this session.
Traceback (most recent call last):
  File "F:\GitHub\TagStudio/tagstudio\src\core\library\alchemy\library.py", line 641, in add_field_tag
    field.tags = field.tags | {tag}
    ^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 537, in __set__
    self.impl.set(
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 1985, in set
    collections.bulk_replace(
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 804, in bulk_replace
    appender(member, _sa_initiator=initiator)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 1394, in add
    value = __set(self, value, _sa_initiator, NO_KEY)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 1096, in __set
    item = executor.fire_append_event(item, _sa_initiator, key=key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 688, in fire_append_event
    return self.attr.fire_append_event(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 1752, in fire_append_event
    value = fn(state, value, initiator or self._append_token, key=key)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\unitofwork.py", line 70, in append
    sess._save_or_update_state(item_state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 3507, in _save_or_update_state
    self._save_or_update_impl(st_)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 4202, in _save_or_update_impl
    self._update_impl(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 4191, in _update_impl
    self.identity_map.add(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\identity.py", line 195, in add
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Can't attach instance <Tag at 0x21280572300>; another instance with key (<class 'src.core.library.alchemy.models.Tag'>, (1000,), None) is already present in this session.
</pre>
</details>

Error after creating tag "Cat", then searching for "cat". Results show as intended, but a traceback remains:
<details><summary markdown="span">Show Logs</summary>
<pre>
2024-09-06 20:12:36 [info     ] searching library              filter=FilterState(page_index=0, page_size=500, search_mode=<SearchMode.AND: 0>, tag='cat', tag_id=None, id=None, path=None, name=None, query='cat') query_full=SELECT entries.id, entries.path 
FROM entries JOIN tag_box_fields ON entries.id = tag_box_fields.entry_id JOIN tag_fields AS tag_fields_1 ON tag_box_fields.id = tag_fields_1.field_id JOIN tags ON tags.id = tag_fields_1.tag_id
WHERE (lower(tags.name) LIKE lower('cat') OR lower(tags.shorthand) LIKE lower('cat')) AND lower(entries.path) NOT LIKE lower('%..json,.xmp,.aae')
 LIMIT 500 OFFSET 0
2024-09-06 20:12:36 [info     ] items to render                count=8
2024-09-06 20:12:36 [info     ] toggle_item_tag                entry_id=4 field_key=TAGS_META tag_id=1 toggle_value=True
2024-09-06 20:12:36 [error    ] Can't attach instance <Tag at 0x212c2484380>; another instance with key (<class 'src.core.library.alchemy.models.Tag'>, (1,), None) is already present in this session.
Traceback (most recent call last):
  File "F:\GitHub\TagStudio/tagstudio\src\core\library\alchemy\library.py", line 641, in add_field_tag
    field.tags = field.tags | {tag}
    ^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 537, in __set__
    self.impl.set(
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 1985, in set
    collections.bulk_replace(
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 804, in bulk_replace
    appender(member, _sa_initiator=initiator)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 1394, in add
    value = __set(self, value, _sa_initiator, NO_KEY)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 1096, in __set
    item = executor.fire_append_event(item, _sa_initiator, key=key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\collections.py", line 688, in fire_append_event
    return self.attr.fire_append_event(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\attributes.py", line 1752, in fire_append_event
    value = fn(state, value, initiator or self._append_token, key=key)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\unitofwork.py", line 70, in append
    sess._save_or_update_state(item_state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 3501, in _save_or_update_state
    self._save_or_update_impl(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 4202, in _save_or_update_impl
    self._update_impl(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\session.py", line 4191, in _update_impl
    self.identity_map.add(state)
  File "f:\GitHub\TagStudio\.venv\Lib\site-packages\sqlalchemy\orm\identity.py", line 195, in add
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Can't attach instance <Tag at 0x212c2484380>; another instance with key (<class 'src.core.library.alchemy.models.Tag'>, (1,), None) is already present in this session.
2024-09-06 20:12:36 [info     ] update_widgets                 selected=[2]
</pre>
</details>

When switching between previews of entries with or without Title text, sometimes the text is not updated between switching entries, and also appears heavily shifted to the right. I can't pinpoint the conditions to cause this at the moment.
![image](https://github.com/user-attachments/assets/5ff91ad8-1931-4d6f-aab3-1041149591b2)

Additionally:
- When editing tags, the color is automatically shifted forward by one index - I remember reading a comment in the code about this, just wanted to bring it up again.
- I don't seem to be able to add new fields to entries
- Clicking on tags to search for them via ID no longer works (I wouldn't worry too much about this, since after this PR would be a great time to get rid of the ID search syntax in favor of a better search bar UI anyways)

CyanVoxel commented:
I'm also still fine with pulling this in a state like this, I just wanted to make sure there wasn't anything critically wrong before doing so

yedpodtrzitko commented:
I will go through the notes you posted, and I will also do one more futureproofing change.

yedpodtrzitko commented:
> Error upon loading a SQLite library populated with tags applied to entries:

should be fixed

> 
> I created a tag "Cat", then a tag "Kitten" with "Cat" as a subtag (didn't expect it to have any lasting effect), then applied "Cat" to an entry under content tags, then attempted to add "Kitten" as well, but failed.

fixed


> Error after creating tag "Cat", then searching for "cat". Results show as intended, but a traceback remains:

should be also fixed

> When switching between previews of entries with or without Title text, sometimes the text is not updated between switching entries, and also appears heavily shifted to the right. I can't pinpoint the conditions to cause this at the moment.

🤔 I dont think this could be caused by the db backend, and I cant even replicate that, but I'll keep an eye on it

> Additionally:
> 
>     * When editing tags, the color is automatically shifted forward by one index - I remember reading a comment in the code about this, just wanted to bring it up again.

fixed

> 
>     * I don't seem to be able to add new fields to entries

fixed

> 
>     * Clicking on tags to search for them via ID no longer works (I wouldn't worry too much about this, since after this PR would be a great time to get rid of the ID search syntax in favor of a better search bar UI anyways)

fixed

CyanVoxel commented:
I'll check out these changes soon and get back to you 👍 

In the meantime, I noticed that the `apprun.yaml` workflow was removed. Was that intentional?

yedpodtrzitko commented:
> In the meantime, I noticed that the `apprun.yaml` workflow was removed. Was that intentional?

Yes, that's on purpose. It happened that there was some error when running the app a few times, but `apprun` job was still green. I replaced it with in `tests/qt`. Each of these is testing a different component of the app, so the overall coverage is much better. It's also more granular so when a test for a specific component fails, you know where to start looking.

CyanVoxel commented:
Thank you again SO MUCH for all of your hard work on this!! 🎉

Tishj commented:
To answer a TODO that was left in the code by this PR:

```py
        if result == 3:  # TODO - what is this magic number?
            callback()
```
The answer is here, it's an enum: <https://doc.qt.io/qt-6/qmessagebox.html#ButtonRole-enum>
